### PR TITLE
[ML] provide nicer errors on _infer with missing model

### PR DESCRIPTION
**Pull request by benwtrent**
When a model doesn't exist, we should indicate that when calling `_infer`.

### CODE REVIEW

**Review by davidkyle — APPROVED** [review]
LGTM

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInferTrainedModelDeploymentAction.java [hunk 1]
```diff
@@ -58,28 +65,39 @@ protected void doExecute(
...
-            String message = "Cannot perform requested action because deployment [" + deploymentId + "] is not yet running on any node";
-            listener.onFailure(ExceptionsHelper.conflictStatusException(message));
-            return;
-        }
-        // TODO Do better routing for inference calls
-        int nodeIndex = Randomness.get().nextInt(randomRunningNode.length);
-        request.setNodes(randomRunningNode[nodeIndex]);
-        super.doExecute(task, request, listener);
+        final String deploymentId = request.getDeploymentId();
+        provider.getTrainedModel(deploymentId, GetTrainedModelsAction.Includes.empty(), ActionListener.wrap(config -> {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInferTrainedModelDeploymentAction.java:69]: This is quite expensive to be doing on each inference request. Could we make this work by just reading the allocation metadata?

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInferTrainedModelDeploymentAction.java:69]: > Could we make this work by just reading the allocation metadata?

I agree, if `_infer` is to be used in a hot-path, we shouldn't grab the model everytime. The trick is, using the deployment metadata only tells us if the deployment exists or not, not if the model exists or not. 

I will think on this again, we could simply update the error to be a nicer 404 message indicating the deployment and/or the model doesn't exist.

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInferTrainedModelDeploymentAction.java:69]: @dimitris-athanasiou how about this, if the deployment is missing, we THEN check if the model exists or not to provide a nicer message?

This way, the `_infer` path is fast when the deployment exists.

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInferTrainedModelDeploymentAction.java:69]: Yeah, I was thinking we could do this too. We could be potentially making ingest slower but only scenarios where there is a clear misconfiguration for the user's pipelines.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Pull request merged**